---
title: "OmicsON"
author: "Monika Piwowar, Wiktor Jurkowski, Ewa Gajda, Cezary Turek"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{OmicsON}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---



```{r global_options, include=TRUE}
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE, error = TRUE)
```

# Table of Contents
i. [About OmicsON](#about-omicson)
i. [OmicsON workflow](#omicson-workflow)
  * 1. [Set up mapping files](set-up-mapping-files)
  * 2. [Data input](#input-data)
  * 3. [Decorate data by Reactome data](#uecorate-data-by-reactome-data)
  * 4. [Decorate data by STRING data](#decorate-data-by-string-data)
  * 5. [Define group](#define-goups)
  * 6. [Lalalala lala](#define-groups-of-genes-and-metabolites)
  * 7. [Multivariate Statistical Analysis](#statistical-analysis)
    * a. [CCA - Canonical Correlation Analysis](#cca---canonical-correlation-analysis)
    * b. [PLS - Partial Least Squares Regression](#pls---partial-least-squares-regression)
ii. [Methods in deatils](#methods-in-deatils)
  * 1. [clasterUsingOntology()](#clasterusingontology())


## About OmicsON
OmicsON provides knowedge driven data regularisation to facilitate multivariate analysis of 'omics' data.

## OmicsON Workflow
Below you will find OmicsON workflow described step by step. By following these steps and using examplary data included in this package user should get the same results as shown in this vignette. 


### Set up mapping files
It is important to set up OmicsON mapping. You can set it up by invoking function presented below. ChEBI to Reactome, Ensembl to Reactome and UniProt to Reactome mappings are required.

```{r, results = 'asis'}
    OmicsON::setUpReactomeMapping(ChEBI2ReactomeFileURL = "https://reactome.org/download/current/ChEBI2Reactome.txt", 
                                  Ensembl2ReactomeFileURL = "https://reactome.org/download/current/Ensembl2Reactome.txt", 
                                  UniProt2ReactomeFileURL = "https://reactome.org/download/current/UniProt2Reactome.txt")
```


### Data input
After setting up mapping, to start work with OmicsON you need to provide two 'omics' data sets in data frame form. Data frame can be created from files as in \extdata directory under package insallation directory. This files are in tab delimited files with headers (colnames and rownames). Files are named `nm-transcriptomics.txt` and `nm-lipidomics.txt`.

Below you can find first few lines of files presented in form of data frame. As you can see all of them have heades and colnames. To find them localization please run `find.package("OmicsON")` after package loading `library(OmicsON)`.
You can use below snippet of code to load mentioned files into R environment:

```{r, results = 'asis'}
    pathToFileWithLipidomicsData <- system.file(package="OmicsON", "extdata", "nm-lipidomics.txt")
    lipidomicsInputData <- read.table(pathToFileWithLipidomicsData, header = TRUE)
    lipidomicsInputDf <- head(lipidomicsInputData, 6)
    knitr::kable(lipidomicsInputDf[1:7], caption = "Lipidomisc data")
    
    pathToFileWithTranscriptomicsData <- system.file(package="OmicsON", "extdata", "nm-transcriptomics.txt")
    transcriptomicsInputData <- read.table(pathToFileWithTranscriptomicsData, header = TRUE)
    transcriptomicsInputDf <- head(transcriptomicsInputData, 6)
    knitr::kable(transcriptomicsInputDf[1:7], caption = "Transcriptomics data")
```


### Decorate data by Reactome data

As soon as you set up OmincsON and input data are loaded, you are ready to decorate data by data presented in Reactome database. It is done by searching of ontologically related molecules present in Reactome's pathways. For further processes in that vignette we are using only few rows from input data. Rows are chosen to present possible border cases in a short time. CHEBI:28875 has representation in Reactome pathways. CHEBI:73705 does not have, but we can find  representatn of that group by ontology in ChEBI database - CHEBI:46761. CHEBI:35465 is not interesting case, no representation and no representants in ChEBI, so mapping and decoration is empty for that id.

```{r, echo=TRUE, results='asis'}
    dataToVignetteEvaluation <- lipidomicsInputData[lipidomicsInputData[,"ChEBI"] %in% c("CHEBI:28875","CHEBI:73705", "CHEBI:35465"),]
    decoratedByReactome <- OmicsON::decorateByReactomeData(chebiMoleculesDf = dataToVignetteEvaluation,
                                chebiIdsColumnName = "ChEBI", organismTaxonomyId = '9606')
    knitr::kable(decoratedByReactome)
```


What algorithm is behind the ontologies mapping? The first two columns of result table represent mapping of all small molecules to respective parents and children of ChEBI ontology: "root" denotes source IDs; if ID already exists in Reactome ontologyId column has the same value, if not but there is an alternative in the form of child or parent in ChEBI ontology, OmicsON put its id under "ontologyId" column. If we can not find root id and can not find any parent or children for it, OmicsON leave it empty.

Full result's data frame contains respectively:

- root - ChEBIs ids given by user,
- ontologyId - ChEBI ids used in the calculation, it is taken from ChEBI ontology base on root,
- ensembleIds - List including vector of Ensemble's Ids,
- uniProtIds - List including vector of UniProt's Ids,
- reactomeIds - List including vector of pathway's ids from Reactome DB,
- genesSymbolsFromEnsemble - List including vector of gen's symbols from Reactome DB base on pathway and Ensemble's Ids,
- genesSymbolsFromUniProt - List including vector of gen's symbols from Reactome DB base on pathway and UniProt's Ids,


### Decorate data by STRING data

When you have results from Reactome step, then you are ready to use decoration by STRING DB. In this part you search for any extra interactions of gens which you find in Reactome. STRING calls them neighbours. To do it just put results achived from Reactome's decoration step to `OmicsON::decorateByStringDbData` method and set `listOfEnsembleIdColumnName` attribute to proper value - `ensembleIds` or `uniProtIds`. This function produce data frame. Below we have presented two data frames respectively for `ensembleIds` and `uniProtIds`:

```{r, echo=TRUE, results='asis'}
    decoratedByStringBaseOnEnsembleIds <- OmicsON::decorateByStringDbData(chebiIdsToReactomePathways = decoratedByReactome, 
                                                               listOfEnsembleIdColumnName = 'ensembleIds')
    decoratedByStringBaseOnUniProtIds <- OmicsON::decorateByStringDbData(chebiIdsToReactomePathways = decoratedByReactome, 
                                                                         listOfEnsembleIdColumnName = 'uniProtIds')
```

Data frame returned from this method introduces three new columns. Them are, respectively:

- stringIds - List including vector of all STRING's ids used in computations. 
- stringGenesSymbolsExpand - List including vector of all neighbours find in STRING database.
- stringGenesSymbolsNarrow - List including vector of intersection of all neighbours per id from set of ids used in search.


How to traverse through data in presented data frames? I will present it on example. Moste importan is to understand that many cells include list of vectors! Example will show also what mapping to Ensemble can give different results than mapping to UniProt:
```{r, echo=FALSE, results='asis'}
    decoratedByStringBaseOnEnsembleIds[1, "root"]
    decoratedByStringBaseOnEnsembleIds[1, "ensembleIds"][[1]]
    decoratedByStringBaseOnEnsembleIds[1, "stringGenesSymbolsNarrow"][[1]]
    decoratedByStringBaseOnUniProtIds[1, "root"]
    decoratedByStringBaseOnUniProtIds[1, "uniProtIds"][[1]]
    decoratedByStringBaseOnUniProtIds[1, "stringGenesSymbolsNarrow"][[1]]
```


### Functional Ineractions DF
You can create a functional interactions data frame by using this method:

```{r, echo=TRUE, results='asis'}
    functionalInteractions <- OmicsON::createFunctionalInteractionsDataFrame(decoratedByReactome)
```

```{r, echo=FALSE, results='asis'}
    knitr::kable(head(functionalInteractions, 6))
```


### Statistical Analysis

OmicsON provides two statistical methods to analyse those data:

- CCA - Canonical Correlation Analysis
- PLS - Partial Least Squares Regression


#### CCA - Canonical Correlation Analysis

Calculate CCA by hand:


```{r, echo=TRUE, results='asis'}
    pathToExampleFileWithXData <- paste(find.package("OmicsON"),"/example/nm-transcriptomics.txt", sep = "")
    pathToExampleFileWithYData <- paste(find.package("OmicsON"),"/example/nm-lipidomics.txt", sep = "")

    XDF <- read.table(pathToExampleFileWithXData, header = TRUE);
    YDF <- read.table(pathToExampleFileWithYData, header = TRUE);

    ccaResults1 <- OmicsON::makeCanonicalCorrelationAnalysis(
        xNamesVector = joinRecatomeTrans,
        yNamesVector = joinLip,
            XDataFrame = XDF,
            YDataFrame = YDF)
```


```{r, fig.show='hold', fig.width=6, fig.height=6}
    OmicsON::plotCanonicalCorrelationAnalysisResults(ccaResults = ccaResults1)
```

Calculate CCA on groups:

```{r, echo=TRUE, results='hide'}
    mccReactome <- OmicsON::makeCCAOnGroups(
        groupsDefinitionDF = groups, 
        mappingDF = chebiIdsToReactomePathwaysWithRoot, 
        groupsDataDF = YDF, 
        mappingDataDF = XDF)

``` 

Inside the makeCCAOnGroups functions the permutatuin test is automaticly done. To do it yourself on your data do:

```{r, echo=FALSE, results='hide'}
    permutationTestsResults <- OmicsON::makePermutationTestOnCCA(
        XDataFrame = XDF, 
        YDataFrame = YDF, 17, 2, 50, 
        countedCCA = ccaResults1)
```

To take data from grouped CCA you should know a little the structure, it is like that:
```{r, echo=FALSE, results='hide'}
mccReactome$Molecules[1]
mccReactome$right[[1]]
mccReactome$left[[1]]
mccReactome$ccaResults[[1]]
mccReactome$ccaPermutationTestResults[[1]]
```

```{r, fig.show='hold', fig.width=6, fig.height=6}
    OmicsON::plotCanonicalCorrelationAnalysisResults(ccaResults = mccReactome$ccaResults[[1]])
```
#### PLS - Partial Least Squares Regression

```{r, echo=FALSE, results='hide'}
    PLSResult1 <- OmicsON::makePartialLeastSquaresRegression(
        joinRecatomeTrans,
        joinLip,
        XDataFrame = XDF,
        YDataFrame = YDF)
```

```{r, fig.show='hold', fig.width=6, fig.height=6}
    OmicsON::plotRmsepForPLS(PLSResult1$training)
```  

```{r, fig.show='hold', fig.width=6, fig.height=6}
    OmicsON::plotRegression(PLSResult1$training, ncompValue = 10)
```

The same but on groups defined by user.
```{r, echo=TRUE, results='asis'}
    groupPlsReactome <- OmicsON::makePLSOnGroups(
        groupsDefinitionDF = groups, 
        mappingDF = chebiIdsToReactomePathwaysWithRoot, 
        groupsDataDF = YDF, 
        mappingDataDF = XDF)

```

How to take counted values from results:
```{r, echo=FALSE, results='hide'}
groupPlsReactome$Molecules[1]
groupPlsReactome$right[[1]]
groupPlsReactome$left[[1]]
groupPlsReactome$plsResults[[1]]
groupPlsReactome$plsPermutationTestResults[[1]]
```
How to use it to plot:
```{r, fig.show='hold', fig.width=6, fig.height=6, echo=TRUE}
    OmicsON::plotRmsepForPLS(groupPlsReactome$plsResults[[1]]$training)
```
```{r, fig.show='hold', fig.width=6, fig.height=6, echo=TRUE}
    OmicsON::plotRegression(groupPlsReactome$plsResults[[1]]$training, ncompValue = 10)
```



